include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h

```diff
@@ -314,8 +314,9 @@ __parallel_transform_reduce(_ExecutionPolicy&& __exec, _ReduceOp __reduce_op, _T
     // TODO: find a way to generalize getting of reliable work-group size
     ::std::size_t __work_group_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
     // change __work_group_size according to local memory limit
+    // Pessimistically double the memory requirement to take into account memory used by compiled kernel
     __work_group_size = oneapi::dpl::__internal::__max_local_allocation_size(::std::forward<_ExecutionPolicy>(__exec),
-                                                                             sizeof(_Tp), __work_group_size);
+                                                                             sizeof(_Tp) * 2, __work_group_size);
```

Comment:
May be better to use in this place ```/ 2``` as in other places ?

Reply:
Reduce is a little different here because instead of using `get_info<sycl::info::device::local_mem_size>()` directly like sort and scan do, it calls `__max_local_allocation_size` which internally uses  `get_info<sycl::info::device::local_mem_size>()`. So instead, we double the memory requirement as opposed to halving the available memory.

Reply:
I think we achieve the desired effect here, but are abusing the arguments to the function. 
The argument is `__type_size`, and we are doubling the true type size to end up with something which then doesn't exceed half the memory size.  
We could consider adding an additional optional float parameter defaulted to 1 which scales the memory returned from the device query prior to using it.  I **would** use a float here and then cast back to integer because otherwise you severely limit the expressiveness of that option.

Reply:
I have a patch already that refactors this function, with the semantics of this parameter to be "the amount of SLM needed per work item". This change seems to be OK with that new semantics. As this case requires O(workgroup size) of local memory and computes the group size based on local memory size, doubling the requirement essentially reduces the effective WG size (and the real memory needed) twice. If that size is still at least 512, then the real kernels will use at most half of the available memory.

The bottom line is that I believe this change is good and achieves the goal we want.

Reply:
Works for me.  In any case, the comment above the line of code should resolve any question of the intention of the call.